### ISSUE TITLE
Hash#dup drops class info

### ISSUE THREAD

**Issue by i110** [mruby/mruby#4030]
```ruby
class MyHash < Hash
end
puts MyHash.new.dup.class
```
says `Hash`, but I expected `MyHash`. OTOH

```ruby
class MyArray < Array
end
puts MyArray.new.dup.class
```

says `MyArray`. Is this a bug of Hash?

**Comment by i110** [comment]
I also found another problem (but related one I guess)

```ruby
h = Hash.new {|h, k| "110"}
h = h.merge({})
p h[:notexist]
```

In CRuby, above code says `110`, while mruby says nil

**Comment by i110** [comment]
@matz Sorry for confusion, but the first issue is not addressed yet by #4031. Could you reopen this?

**Comment by matz** [comment]
To avoid this kind of confusion, please separate issues next time.
